File: app/src/main/java/com/github/malitsplus/shizurunotes/ui/base/BackgroundSpan.java
package com.github.malitsplus.shizurunotes.ui.base;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.text.style.ReplacementSpan;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.github.malitsplus.shizurunotes.R;
import com.github.malitsplus.shizurunotes.common.ResourceManager;

public class BackgroundSpan extends ReplacementSpan {

    public static final int BORDER_RECT = 1;
    public static final int BACKGROUND_RECT = 2;

    private int type;

    public BackgroundSpan(int type) {
        this.type = type;
    }

    @Override
    public int getSize(@NonNull Paint paint, CharSequence text, int start, int end, @Nullable Paint.FontMetricsInt fontMetricsInt) {
        return (int) paint.measureText(text, start, end);
    }

    @Override
    public void draw(@NonNull Canvas canvas, CharSequence text, int start, int end, float x, int top, int y, int bottom, @NonNull Paint paint) {

        //先填充文字
        paint.setStyle(Paint.Style.FILL);
        canvas.drawText(text, start, end, x, y, paint);

        //设置边框粗细
        paint.setStrokeWidth(3.0f);
        //设置绘制矩形范围
        RectF rectF = new RectF(x, top + 5, x + measureText(paint, text, start, end), bottom - 5);

        switch (type){
            case BORDER_RECT:
                drawBorderRect(canvas, paint, rectF);
                break;
            case BACKGROUND_RECT:
                drawBackgroundRect(canvas, paint, rectF);
                break;
        }
    }

    private float measureText(Paint paint, CharSequence text, int start, int end) {
        return paint.measureText(text, start, end);
    }

    private void drawBorderRect(Canvas canvas, Paint paint, RectF rectF) {
        int cornerRadius = 10;

        //画无边框的纯填充色方框背景
        paint.setColor(Color.TRANSPARENT);
        canvas.drawRoundRect(rectF, cornerRadius, cornerRadius, paint);

        //画边框线
        paint.setStyle(Paint.Style.STROKE);
        paint.setColor(Color.GRAY);
        paint.setAntiAlias(true);
        canvas.drawRoundRect(rectF, cornerRadius, cornerRadius, paint);
    }

    private void drawBackgroundRect(Canvas canvas, Paint paint, RectF rectF) {
        int cornerRadius = 10;

        //画无边框的纯填充色方框背景
        paint.setColor(ResourceManager.Companion.get().getColor(R.color.red_200_beta));
        canvas.drawRoundRect(rectF, cornerRadius, cornerRadius, paint);

        //画边框线
        paint.setStyle(Paint.Style.STROKE);
        paint.setColor(Color.GRAY);
        paint.setAntiAlias(true);
        canvas.drawRoundRect(rectF, cornerRadius, cornerRadius, paint);
    }
}

File: app/src/main/java/com/github/nyanfantasia/shizurunotes/ui/base/BackgroundSpan.kt
package com.github.nyanfantasia.shizurunotes.ui.base

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.text.style.ReplacementSpan
import com.github.nyanfantasia.shizurunotes.R
import com.github.nyanfantasia.shizurunotes.common.ResourceManager.Companion.get

class BackgroundSpan(private val type: Int) : ReplacementSpan() {
    override fun getSize(
        paint: Paint,
        text: CharSequence,
        start: Int,
        end: Int,
        fontMetricsInt: Paint.FontMetricsInt?
    ): Int {
        return paint.measureText(text, start, end).toInt()
    }

    override fun draw(
        canvas: Canvas,
        text: CharSequence,
        start: Int,
        end: Int,
        x: Float,
        top: Int,
        y: Int,
        bottom: Int,
        paint: Paint
    ) {

        //先填充文字
        paint.style = Paint.Style.FILL
        canvas.drawText(text, start, end, x, y.toFloat(), paint)

        //设置边框粗细
        paint.strokeWidth = 3.0f
        //设置绘制矩形范围
        val rectF = RectF(
            x,
            (top + 5).toFloat(),
            x + measureText(paint, text, start, end),
            (bottom - 5).toFloat()
        )
        when (type) {
            BORDER_RECT -> drawBorderRect(canvas, paint, rectF)
            BACKGROUND_RECT -> drawBackgroundRect(canvas, paint, rectF)
        }
    }

    private fun measureText(paint: Paint, text: CharSequence, start: Int, end: Int): Float {
        return paint.measureText(text, start, end)
    }

    private fun drawBorderRect(canvas: Canvas, paint: Paint, rectF: RectF) {
        val cornerRadius = 10

        //画无边框的纯填充色方框背景
        paint.color = Color.TRANSPARENT
        canvas.drawRoundRect(rectF, cornerRadius.toFloat(), cornerRadius.toFloat(), paint)

        //画边框线
        paint.style = Paint.Style.STROKE
        paint.color = Color.GRAY
        paint.isAntiAlias = true
        canvas.drawRoundRect(rectF, cornerRadius.toFloat(), cornerRadius.toFloat(), paint)
    }

    private fun drawBackgroundRect(canvas: Canvas, paint: Paint, rectF: RectF) {
        val cornerRadius = 10

        //画无边框的纯填充色方框背景
        paint.color = get()
            .getColor(R.color.red_200_beta)
        canvas.drawRoundRect(rectF, cornerRadius.toFloat(), cornerRadius.toFloat(), paint)

        //画边框线
        paint.style = Paint.Style.STROKE
        paint.color = Color.GRAY
        paint.isAntiAlias = true
        canvas.drawRoundRect(rectF, cornerRadius.toFloat(), cornerRadius.toFloat(), paint)
    }

    companion object {
        const val BORDER_RECT = 1
        const val BACKGROUND_RECT = 2
    }
}